File: Data Structures & Algorithms/coin-change/submission-0.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        
        dp=[[0]*(len(coins)+1) for _ in range(amount+1)]
        
        for total in range(1,len(dp)):
            for c in range(1,len(dp[0])):
                if total>=coins[c-1] and (dp[total-coins[c-1]][c]!=0 or total-coins[c-1]==0) :
                    
                    temp=dp[total-coins[c-1]][c]+1
                    if dp[total][c-1]!=0:
                        dp[total][c]=min(dp[total][c-1],temp)
                    else:
                        dp[total][c]=temp
                    
                else:
                    dp[total][c]=dp[total][c-1]
        
        
        if dp[amount][len(coins)]==0 and amount!=0:
            return -1
        return dp[amount][len(coins)]

File: Data Structures & Algorithms/coin-change-ii/submission-0.java
class Solution {
    int coinChange(int i,int amount,int coins[],int dp[][])
    {
        if( i==coins.length )
            return 0;
         if(amount==0)
            return 1;

        if( amount<0 || amount<coins[i])
            return 0;
       
        if(dp[i][amount]!=-1)
            return dp[i][amount];

        return dp[i][amount]=coinChange(i,amount-coins[i],coins,dp)+coinChange(i+1,amount,coins,dp);

    }
    public int change(int amount, int[] coins) {
    int dp[][] = new int[coins.length][amount+1];
    Arrays.sort(coins);
    for(int i=0;i<coins.length;i++)
        for(int j=0;j<amount+1;j++)
            dp[i][j]=-1;
    return coinChange(0,amount,coins,dp);

    }
}